### ISSUE TITLE
[Bug]: Cannot view/restore versions of Objects with Classification Stores

### ISSUE THREAD

**Issue by NiklasBr** [pimcore/pimcore#10661]
### Expected behavior

When clicking on the *Versions* tab on DataObjects it should be possible to view and restore from versions created prior to Pimcore X which contains Classification Stores.

### Actual behavior

```log
[2021-10-20T09:56:31.636336+02:00] php.CRITICAL: Uncaught Error: 
Cannot assign __PHP_Incomplete_Class to property Pimcore\Model\DataObject\Classificationstore::$object of type ?Pimcore\Model\DataObject\Concrete {"exception":"[object] (TypeError(code: 0): 
Cannot assign __PHP_Incomplete_Class to property Pimcore\\Model\\DataObject\\Classificationstore::$object of type ?Pimcore\\Model\\DataObject\\Concrete at /var/www/pimcore/vendor/pimcore/pimcore/lib/Tool/Serialize.php:43)"} []
[2021-10-20T09:56:31.636655+02:00] request.CRITICAL: Uncaught PHP Exception TypeError: 
"Cannot assign __PHP_Incomplete_Class to property Pimcore\Model\DataObject\Classificationstore::$object of type ?Pimcore\Model\DataObject\Concrete" at /var/www/pimcore/vendor/pimcore/pimcore/lib/Tool/Serialize.php line 43 {"exception":"[object] (TypeError(code: 0): 
Cannot assign __PHP_Incomplete_Class to property Pimcore\\Model\\DataObject\\Classificationstore::$object of type ?Pimcore\\Model\\DataObject\\Concrete at /var/www/pimcore/vendor/pimcore/pimcore/lib/Tool/Serialize.php:43)"} []

```

![image](https://user-images.githubusercontent.com/279826/138054399-4607b73c-77ba-4c57-9a59-c59496027504.png)

### Steps to reproduce

Save a version in Pimcore 6.9.6
Upgrade to Pimcore 10.1+
Try to restore the object

*edit* example version file (renamed with `.txt` to allow upload here on Github)
[255796.txt](https://github.com/pimcore/pimcore/files/7379969/255796.txt)

**Comment by NiklasBr** [comment]
@dvesh3 it's really hard to do since the debug flag was removed from System settings (unclear why, do you know?), so I cannot do it quickly. Will attempt to do it elsewhere.

But I found *another issue* with versions and Classification Stores just now, versions created in Pimcore 10.1.5 cannot be viewed in 10.1.5 either:

![image](https://user-images.githubusercontent.com/279826/138069770-d60d8482-4a69-427b-9d26-5fa7220b8a61.png)
```trace
Twig\Error\RuntimeError:
Key "1" for array with keys "20" does not exist.

  at vendor/pimcore/pimcore/bundles/AdminBundle/Resources/views/Admin/DataObject/DataObject/previewVersion.html.twig:117
  at twig_get_attribute(object(Environment), object(Source), array(true), true, array(), 'array', false, false, false, 117)
     (/var/pim/cache/qa/twig/a4/a48c171ea5726a02af83e7140cff98a449fcbfd0708c7f7e7b241cf732e5cd9a.php:428)
  at __TwigTemplate_51a0e00a73cd2ee11bb36ddcd20e40db45bc92893a65eed69f43dc51e7b5be98->{closure}(true, 20)
  at array_filter(array(true), object(Closure), 1)
     (vendor/twig/twig/src/Extension/CoreExtension.php:1594)
  at twig_array_filter(object(Environment), array(true), object(Closure))
     (/var/pim/cache/qa/twig/a4/a48c171ea5726a02af83e7140cff98a449fcbfd0708c7f7e7b241cf732e5cd9a.php:428)
  at __TwigTemplate_51a0e00a73cd2ee11bb36ddcd20e40db45bc92893a65eed69f43dc51e7b5be98->doDisplay(array('object' => object(Product), 'versionNote' => '', 'validLanguages' => array('en', 'da', 'nb', 'sv'), 'app' => object(AppVariable), 'container' => object(App_KernelQaDebugContainer), 'pimcore_csrf' => object(CsrfProtectionHandler), 'document' => null, 'editmode' => false, 'csrfToken' => 'beff836a100a13dc7704d9447d18fedd9e69ba82', 'fields' => array('productNumberCompany' => object(Input), 'salesUnit' => object(Select), 'salesAmount' => object(Select), 'priceUnit' => object(Select), 'ETIMClass' => object(Input), 'category' => object(ManyToManyObjectRelation), 'localizedfields' => object(Localizedfields), 'productData' => object(Table), 'technicalData' => object(Table), 'tableBlock' => object(Block), 'Count' => object(Input), 'EAN' => object(Input), 'Standard' => object(ManyToManyObjectRelation), 'ValueType' => object(Input), 'blockNumber' => object(Input), 'ROHS' => object(Input), 'REACH' => object(Input), 'alem09' => object(Input), 'dataStoreEnergy' => object(Classificationstore), 'dataStore' => object(Classificationstore), 'dataStoreInfrastructure' => object(Classificationstore), 'dataStoreInstallation' => object(Classificationstore), 'netLength' => object(QuantityValue), 'netHeight' => object(QuantityValue), 'netWidth' => object(QuantityValue), 'netWeight' => object(QuantityValue), 'productOwner' => object(User), 'productImages' => object(ImageGallery), 'environmentalImages' => object(ImageGallery), 'productVideos' => object(AdvancedManyToManyRelation), 'instructionsForUse' => object(AdvancedManyToManyRelation), 'dimensionalSketches' => object(AdvancedManyToManyRelation), 'installationInstructions' => object(AdvancedManyToManyRelation), 'productIllustrations3d' => object(AdvancedManyToManyRelation), 'otherDocuments' => object(AdvancedManyToManyRelation), 'images' => object(ImageGallery), 'files' => object(AdvancedManyToManyRelation)), 'definition' => object(Classificationstore), 'fieldName' => 'dataStoreEnergy', '_parent' => array('object' => object(Product), 'versionNote' => '', 'validLanguages' => array('en', 'da', 'nb', 'sv'), 'app' => object(AppVariable), 'container' => object(App_KernelQaDebugContainer), 'pimcore_csrf' => object(CsrfProtectionHandler), 'document' => null, 'editmode' => false, 'csrfToken' => 'beff836a100a13dc7704d9447d18fedd9e69ba82', 'fields' => array('productNumberCompany' => object(Input), 'salesUnit' => object(Select), 'salesAmount' => object(Select), 'priceUnit' => object(Select), 'ETIMClass' => object(Input), 'category' => object(ManyToManyObjectRelation), 'localizedfields' => object(Localizedfields), 'productData' => object(Table), 'technicalData' => object(Table), 'tableBlock' => object(Block), 'Count' => object(Input), 'EAN' => object(Input), 'Standard' => object(ManyToManyObjectRelation), 'ValueType' => object(Input), 'blockNumber' => object(Input), 'ROHS' => object(Input), 'REACH' => object(Input), 'alem09' => object(Input), 'dataStoreEnergy' => object(Classificationstore), 'dataStore' => object(Classificationstore), 'dataStoreInfrastructure' => object(Classificationstore), 'dataStoreInstallation' => object(Classificationstore), 'netLength' => object(QuantityValue), 'netHeight' => object(QuantityValue), 'netWidth' => object(QuantityValue), 'netWeight' => object(QuantityValue), 'productOwner' => object(User), 'productImages' => object(ImageGallery), 'environmentalImages' => object(ImageGallery), 'productVideos' => object(AdvancedManyToManyRelation), 'instructionsForUse' => object(AdvancedManyToManyRelation), 'dimensionalSketches' => object(AdvancedManyToManyRelation), 'installationInstructions' => object(AdvancedManyToManyRelation), 'productIllustrations3d' => object(AdvancedManyToManyRelation), 'otherDocuments' => object(AdvancedManyToManyRelation), 'images' => object(ImageGallery), 'files' => object(AdvancedManyToManyRelation)), 'definition' => object(Classificationstore), 'fieldName' => 'dataStoreEnergy', '_parent' => array('object' => object(Product), 'versionNote' => '', 'validLanguages' => array('en', 'da', 'nb', 'sv'), 'app' => object(AppVariable), 'container' => object(App_KernelQaDebugContainer), 'pimcore_csrf' => object(CsrfProtectionHandler), 'document' => null, 'editmode' => false, 'csrfToken' => 'beff836a100a13dc7704d9447d18fedd9e69ba82', 'fields' => array('productNumberCompany' => object(Input), 'salesUnit' => object(Select), 'salesAmount' => object(Select), 'priceUnit' => object(Select), 'ETIMClass' => object(Input), 'category' => object(ManyToManyObjectRelation), 'localizedfields' => object(Localizedfields), 'productData' => object(Table), 'technicalData' => object(Table), 'tableBlock' => object(Block), 'Count' => object(Input), 'EAN' => object(Input), 'Standard' => object(ManyToManyObjectRelation), 'ValueType' => object(Input), 'blockNumber' => object(Input), 'ROHS' => object(Input), 'REACH' => object(Input), 'alem09' => object(Input), 'dataStoreEnergy' => object(Classificationstore), 'dataStore' => object(Classificationstore), 'dataStoreInfrastructure' => object(Classificationstore), 'dataStoreInstallation' => object(Classificationstore), 'netLength' => object(QuantityValue), 'netHeight' => object(QuantityValue), 'netWidth' => object(QuantityValue), 'netWeight' => object(QuantityValue), 'productOwner' => object(User), 'productImages' => object(ImageGallery), 'environmentalImages' => object(ImageGallery), 'productVideos' => object(AdvancedManyToManyRelation), 'instructionsForUse' => object(AdvancedManyToManyRelation), 'dimensionalSketches' => object(AdvancedManyToManyRelation), 'installationInstructions' => object(AdvancedManyToManyRelation), 'productIllustrations3d' => object(AdvancedManyToManyRelation), 'otherDocuments' => object(AdvancedManyToManyRelation), 'images' => object(ImageGallery), 'files' => object(AdvancedManyToManyRelation))), '_seq' => array('productNumberCompany' => object(Input), 'salesUnit' => object(Select), 'salesAmount' => object(Select), 'priceUnit' => object(Select), 'ETIMClass' => object(Input), 'category' => object(ManyToManyObjectRelation), 'localizedfields' => object(Localizedfields), 'productData' => object(Table), 'technicalData' => object(Table), 'tableBlock' => object(Block), 'Count' => object(Input), 'EAN' => object(Input), 'Standard' => object(ManyToManyObjectRelation), 'ValueType' => object(Input), 'blockNumber' => object(Input), 'ROHS' => object(Input), 'REACH' => object(Input), 'alem09' => object(Input), 'dataStoreEnergy' => object(Classificationstore), 'dataStore' => object(Classificationstore), 'dataStoreInfrastructure' => object(Classificationstore), 'dataStoreInstallation' => object(Classificationstore), 'netLength' => object(QuantityValue), 'netHeight' => object(QuantityValue), 'netWidth' => object(QuantityValue), 'netWeight' => object(QuantityValue), 'productOwner' => object(User), 'productImages' => object(ImageGallery), 'environmentalImages' => object(ImageGallery), 'productVideos' => object(AdvancedManyToManyRelation), 'instructionsForUse' => object(AdvancedManyToManyRelation), 'dimensionalSketches' => object(AdvancedManyToManyRelation), 'installationInstructions' => object(AdvancedManyToManyRelation), 'productIllustrations3d' => object(AdvancedManyToManyRelation), 'otherDocuments' => object(AdvancedManyToManyRelation), 'images' => object(ImageGallery), 'files' => object(AdvancedManyToManyRelation)), 'loop' => array('parent' => array('object' => object(Product), 'versionNote' => '', 'validLanguages' => array('en', 'da', 'nb', 'sv'), 'app' => object(AppVariable), 'container' => object(App_KernelQaDebugContainer), 'pimcore_csrf' => object(CsrfProtectionHandler), 'document' => null, 'editmode' => false, 'csrfToken' => 'beff836a100a13dc7704d9447d18fedd9e69ba82', 'fields' => array('productNumberCompany' => object(Input), 'salesUnit' => object(Select), 'salesAmount' => object(Select), 'priceUnit' => object(Select), 'ETIMClass' => object(Input), 'category' => object(ManyToManyObjectRelation), 'localizedfields' => object(Localizedfields), 'productData' => object(Table), 'technicalData' => object(Table), 'tableBlock' => object(Block), 'Count' => object(Input), 'EAN' => object(Input), 'Standard' => object(ManyToManyObjectRelation), 'ValueType' => object(Input), 'blockNumber' => object(Input), 'ROHS' => object(Input), 'REACH' => object(Input), 'alem09' => object(Input), 'dataStoreEnergy' => object(Classificationstore), 'dataStore' => object(Classificationstore), 'dataStoreInfrastructure' => object(Classificationstore), 'dataStoreInstallation' => object(Classificationstore), 'netLength' => object(QuantityValue), 'netHeight' => object(QuantityValue), 'netWidth' => object(QuantityValue), 'netWeight' => object(QuantityValue), 'productOwner' => object(User), 'productImages' => object(ImageGallery), 'environmentalImages' => object(ImageGallery), 'productVideos' => object(AdvancedManyToManyRelation), 'instructionsForUse' => object(AdvancedManyToManyRelation), 'dimensionalSketches' => object(AdvancedManyToManyRelation), 'installationInstructions' => object(AdvancedManyToManyRelation), 'productIllustrations3d' => object(AdvancedManyToManyRelation), 'otherDocuments' => object(AdvancedManyToManyRelation), 'images' => object(ImageGallery), 'files' => object(AdvancedManyToManyRelation))), 'index0' => 18, 'index' => 19, 'first' => false, 'revindex0' => 18, 'revindex' => 19, 'length' => 37, 'last' => false), 'storedata' => object(Classificationstore), 'activeGroups' => array(true), 'languages' => array('en', 'da', 'nb', 'sv', 'default')), '_seq' => array('productNumberCompany' => object(Input), 'salesUnit' => object(Select), 'salesAmount' => object(Select), 'priceUnit' => object(Select), 'ETIMClass' => object(Input), 'category' => object(ManyToManyObjectRelation), 'localizedfields' => object(Localizedfields), 'productData' => object(Table), 'technicalData' => object(Table), 'tableBlock' => object(Block), 'Count' => object(Input), 'EAN' => object(Input), 'Standard' => object(ManyToManyObjectRelation), 'ValueType' => object(Input), 'blockNumber' => object(Input), 'ROHS' => object(Input), 'REACH' => object(Input), 'alem09' => object(Input), 'dataStoreEnergy' => object(Classificationstore), 'dataStore' => object(Classificationstore), 'dataStoreInfrastructure' => object(Classificationstore), 'dataStoreInstallation' => object(Classificationstore), 'netLength' => object(QuantityValue), 'netHeight' => object(QuantityValue), 'netWidth' => object(QuantityValue), 'netWeight' => object(QuantityValue), 'productOwner' => object(User), 'productImages' => object(ImageGallery), 'environmentalImages' => object(ImageGallery), 'productVideos' => object(AdvancedManyToManyRelation), 'instructionsForUse' => object(AdvancedManyToManyRelation), 'dimensionalSketches' => object(AdvancedManyToManyRelation), 'installationInstructions' => object(AdvancedManyToManyRelation), 'productIllustrations3d' => object(AdvancedManyToManyRelation), 'otherDocuments' => object(AdvancedManyToManyRelation), 'images' => object(ImageGallery), 'files' => object(AdvancedManyToManyRelation)), 'loop' => array('parent' => array('object' => object(Product), 'versionNote' => '', 'validLanguages' => array('en', 'da', 'nb', 'sv'), 'app' => object(AppVariable), 'container' => object(App_KernelQaDebugContainer), 'pimcore_csrf' => object(CsrfProtectionHandler), 'document' => null, 'editmode' => false, 'csrfToken' => 'beff836a100a13dc7704d9447d18fedd9e69ba82', 'fields' => array('productNumberCompany' => object(Input), 'salesUnit' => object(Select), 'salesAmount' => object(Select), 'priceUnit' => object(Select), 'ETIMClass' => object(Input), 'category' => object(ManyToManyObjectRelation), 'localizedfields' => object(Localizedfields), 'productData' => object(Table), 'technicalData' => object(Table), 'tableBlock' => object(Block), 'Count' => object(Input), 'EAN' => object(Input), 'Standard' => object(ManyToManyObjectRelation), 'ValueType' => object(Input), 'blockNumber' => object(Input), 'ROHS' => object(Input), 'REACH' => object(Input), 'alem09' => object(Input), 'dataStoreEnergy' => object(Classificationstore), 'dataStore' => object(Classificationstore), 'dataStoreInfrastructure' => object(Classificationstore), 'dataStoreInstallation' => object(Classificationstore), 'netLength' => object(QuantityValue), 'netHeight' => object(QuantityValue), 'netWidth' => object(QuantityValue), 'netWeight' => object(QuantityValue), 'productOwner' => object(User), 'productImages' => object(ImageGallery), 'environmentalImages' => object(ImageGallery), 'productVideos' => object(AdvancedManyToManyRelation), 'instructionsForUse' => object(AdvancedManyToManyRelation), 'dimensionalSketches' => object(AdvancedManyToManyRelation), 'installationInstructions' => object(AdvancedManyToManyRelation), 'productIllustrations3d' => object(AdvancedManyToManyRelation), 'otherDocuments' => object(AdvancedManyToManyRelation), 'images' => object(ImageGallery), 'files' => object(AdvancedManyToManyRelation))), 'index0' => 18, 'index' => 19, 'first' => false, 'revindex0' => 18, 'revindex' => 19, 'length' => 37, 'last' => false), 'storedata' => object(Classificationstore), 'activeGroups' => array(true), 'languages' => array('en', 'da', 'nb', 'sv', 'default')), array())
     (vendor/twig/twig/src/Template.php:394)
  at Twig\Template->displayWithErrorHandling(array('object' => object(Product), 'versionNote' => '', 'validLanguages' => array('en', 'da', 'nb', 'sv'), 'app' => object(AppVariable), 'container' => object(App_KernelQaDebugContainer), 'pimcore_csrf' => object(CsrfProtectionHandler), 'document' => null, 'editmode' => false, 'csrfToken' => 'beff836a100a13dc7704d9447d18fedd9e69ba82'), array())
     (vendor/twig/twig/src/Template.php:367)
  at Twig\Template->display(array('object' => object(Product), 'versionNote' => '', 'validLanguages' => array('en', 'da', 'nb', 'sv')))
     (vendor/twig/twig/src/Template.php:379)
  at Twig\Template->render(array('object' => object(Product), 'versionNote' => '', 'validLanguages' => array('en', 'da', 'nb', 'sv')), array())
     (vendor/twig/twig/src/TemplateWrapper.php:40)
  at Twig\TemplateWrapper->render(array('object' => object(Product), 'versionNote' => '', 'validLanguages' => array('en', 'da', 'nb', 'sv')))
     (vendor/twig/twig/src/Environment.php:277)
  at Twig\Environment->render('@PimcoreAdmin/Admin/DataObject/DataObject/previewVersion.html.twig', array('object' => object(Product), 'versionNote' => '', 'validLanguages' => array('en', 'da', 'nb', 'sv')))
     (vendor/symfony/framework-bundle/Controller/AbstractController.php:251)
  at Symfony\Bundle\FrameworkBundle\Controller\AbstractController->renderView('@PimcoreAdmin/Admin/DataObject/DataObject/previewVersion.html.twig', array('object' => object(Product), 'versionNote' => '', 'validLanguages' => array('en', 'da', 'nb', 'sv')))
     (vendor/pimcore/pimcore/lib/Controller/Controller.php:81)
  at Pimcore\Controller\Controller->renderView('@PimcoreAdmin/Admin/DataObject/DataObject/previewVersion.html.twig', array('object' => object(Product), 'versionNote' => '', 'validLanguages' => array('en', 'da', 'nb', 'sv')))
     (vendor/symfony/framework-bundle/Controller/AbstractController.php:259)
  at Symfony\Bundle\FrameworkBundle\Controller\AbstractController->render('@PimcoreAdmin/Admin/DataObject/DataObject/previewVersion.html.twig', array('object' => object(Product), 'versionNote' => '', 'validLanguages' => array('en', 'da', 'nb', 'sv')), null)
     (vendor/pimcore/pimcore/lib/Controller/Controller.php:43)
```

**Comment by dvesh3** [comment]
@NiklasBr I had a look at the version file(uploaded in description). The root cause is that your version files are still referencing to old App class namespace "AppBundle\\.." and Upgrade to Pimcore X requires you change App namespace to "App\\.."

![image](https://user-images.githubusercontent.com/5137917/138078708-4322e249-e0c7-430f-97c6-02889a13ed68.png)

So you have update your version files to have new App namespace, if you would like to use it on Pimcore X.

**Comment by dvesh3** [comment]
> But I found another issue with versions and Classification Stores just now, versions created in Pimcore 10.1.5 cannot be viewed in 10.1.5 either

could you please try this on latest version 10.2.2?

**Comment by NiklasBr** [comment]
Can't try on 10.2 yet because of these changes:
> [Maintenance] Maintenance tasks are now handled with Symfony Messenger. The pimcore:maintenance command will add the maintenance messages to the bus and runs them afterwards immediately from the queue. However it's recommended to setup independent workers that process the queues, by running bin/console messenger:consume pimcore_core pimcore_maintenance (using e.g. Supervisor) and adding --async option to the pimcore:maintenance command that stops the maintenance command to process the queue directly. Details about setting it up for production environments, please check Symfony Messenger Component docs.

This requires changes to our production environment we have not yet have time to test.

**Comment by NiklasBr** [comment]
> So you have update your version files to have new App namespace, if you would like to use it on Pimcore X.

Is there a migration available?

**Comment by dvesh3** [comment]
> Is there a migration available?

Sorry, no migration is available at the moment for migrating versions.

**Comment by NiklasBr** [comment]
When will it be available? This is a pretty big issue at the moment. The namespace had to be changed. I mean, look at your own code which we copied followed according to the guide:

https://github.com/pimcore/demo/blob/58e03bcdb5de55aa981f53b3e9c1b428dbe0fbc4/config/ecommerce/base-ecommerce.yaml#L5-L10

Data losses like this **must** be clearly spelled out in the upgrade documentation!

**Comment by dvesh3** [comment]
> When will it be available? This is a pretty big issue at the moment. The namespace had to be changed. I mean, look at your own code which we copied followed according to the guide:
> 
> https://github.com/pimcore/demo/blob/58e03bcdb5de55aa981f53b3e9c1b428dbe0fbc4/config/ecommerce/base-ecommerce.yaml#L5-L10
> 
> Data losses like this **must** be clearly spelled out in the upgrade documentation!

I have updated the Pimcore X upgrade notes with the point that you should migrate your versions on Pimcore 6.9 to new namespaces if you wish to carry forward these on Pimcore X. https://github.com/pimcore/pimcore/commit/e9ec707bec9ba020a3aced8b321607ce42f83b58

You can adapt and use this migration script https://gist.github.com/weisswurstkanone/a63f733fe58930778f41c695f862724a#file-migrate_element_descriptor_versions-php

Note: It is still possible to live with AppBundle old structure, so it is the developer decision to migrate these version files.

**Comment by NiklasBr** [comment]
The upgrade notes does not make it look like optional:
> You must also upgrade the directory structure to the one shown above [...]
> 7. Move the original PHP source code files from src/AppBundle/*, except bundle specific files (like AppBundle.php and DependencyInjection/), to src/ and update the namespace of each moved file to be App\... (advanced IDEs can do this automatically).

**Comment by dvesh3** [comment]
yes, if you want to use Symfony Flex 😉

**Comment by NiklasBr** [comment]
This does not sound like it is in any way optional:

![image](https://user-images.githubusercontent.com/279826/138237114-e39d3dd5-61df-4c38-b289-4993de11f7e9.png)

**Comment by brusch** [comment]
Yes, sorry, the optional note is missing in the docs, we'll fix that.

**Comment by NiklasBr** [comment]
Will the fix be in 10.2.4 @brusch?

**Comment by brusch** [comment]
@NiklasBr as @dvesh3 mentioned above, I also don't think it makes sense to integrate a generic fix for that. I'm wondering if that would be even possible. 
The only thing we're going to do from our side is updating the docs accordingly.

**Comment by dpfaffenbauer** [comment]
Guess you could still use AppBundle, Symfony doesn't forbid that.

**Comment by brusch** [comment]
That's probably the most easiest fix.

**Comment by NiklasBr** [comment]
> Guess you could still use AppBundle, Symfony doesn't forbid that.

Really, because the linked Symfony docs says that it is a *must*?

> That's probably the most easiest fix.

So what do we do about the tens of thousands of version files now created with `App` as the namespace because we trusted your documentation to be correct? A documentation that *still demands* that we use `App` as namespace despite your comments in this thread.

**Comment by brusch** [comment]
https://symfony.com/doc/current/setup/flex.html#flex-customize-paths

So you could still use `src/AppBundle` along with `src/App`. That's mostly about autoloading, where Composer takes care of.

**Comment by NiklasBr** [comment]
If we do that, what do you expect us to tell our customer about data loss from following your documentation? Lose all version data for the year prior to Pimcore X or all version data for the half year after Pimcore X? This really strains the credibility of Pimcore as a stable and viable platform in their eyes.

**Comment by brusch** [comment]
Another solution that comes to my mind ... use `class_alias()` to map to the old FQCN. 

Create `src/class-aliases.php` with e.g. the following content: 
```php
<?php 

\class_alias(Product::class, 'AppBundle\Model\DataObject\Product');
// ...

```

then include that file in your `composer.json`: 

```json
{
    "autoload": {
        "files": ["src/class-aliases.php"]
    }
}
```

disclaimer: not tested, but should work in theory and easy to test 😉

Regarding your latest answer: 
Well ... we referenced the Symfony docs here to keep things simple, the docs are not 100% clear here, yes agree, but you cannot blame everything on Pimcore, that's just not fair. 
Also your customer is most probably using Pimcore Community Edition ... so, yes ...

**Comment by NiklasBr** [comment]
> Well ... we referenced the Symfony docs here to keep things simple

Which had steps to follow and was clear about the required namespace change:

> You must also upgrade the directory structure to the one shown above [...] update the namespace of each moved file to be App...

> your customer is most probably using Pimcore Community Edition 

Were in the process of acquiring the enterprise license. But at the time, yes.

**Comment by brusch** [comment]
@NiklasBr what do you think about the quick fix above? 😊

**Comment by NiklasBr** [comment]
Promising idea, I will test it.

**Comment by brusch** [comment]
Please let us know if this fixes the issue, we could then update the docs accordingly (a PR of course would be very nice 😉).

**Comment by dpfaffenbauer** [comment]
A solution for this in the long-term would be to change how serialization works. I think it would be much better to use something like Symfony Serializer to serialize and deserialize the data into a json or xml format. that way it does not depend on the resource heavy DeepClone and also not on any Class Names.

**Comment by brusch** [comment]
@dpfaffenbauer absolutely!

**Comment by brusch** [comment]
@NiklasBr did it worked?

**Comment by NiklasBr** [comment]
@brusch class aliases seem to work, cannot get past the Classification Store errors:

```
TypeError:
Cannot assign __PHP_Incomplete_Class to property Pimcore\Model\DataObject\Classificationstore::$object of type ?Pimcore\Model\DataObject\Concrete
at vendor/pimcore/pimcore/lib/Tool/Serialize.php:43
```

**Comment by blankse** [comment]
@NiklasBr Can you retry it with https://github.com/pimcore/pimcore/pull/10878?

**Comment by NiklasBr** [comment]
@blankse I know it was a reason _I_ voted for you for Pimconaut ;) your fix in #10878 appears to work great for viewing objects again in the admin UI when combined with a `class_alias` file loaded using [Composer's Files autoloading functionality](https://getcomposer.org/doc/04-schema.md#files) like this:
```php
use App\Model\DataObject;

class_alias(DataObject\Category::class, 'AppBundle\Model\DataObject\Category');
class_alias(DataObject\Customer::class, 'AppBundle\Model\DataObject\Customer');
class_alias(DataObject\Location::class, 'AppBundle\Model\DataObject\Location');
class_alias(DataObject\News::class, 'AppBundle\Model\DataObject\News');
class_alias(DataObject\Product::class, 'AppBundle\Model\DataObject\Product');
class_alias(DataObject\Standard::class, 'AppBundle\Model\DataObject\Standard');
// ...
```

However, versions created prior to Pimcore X still cannot be restored/published again. Pimcore 10.2.5 throws the following error:
```
HTTP 500 Internal Server Error
Cannot access property starting with "\0"
Exception Error in /var/www/pimcore/vendor/pimcore/pimcore/lib/Tool/Serialize.php (line 105)
```

@brusch Can you update the documentation and clarify which steps are not actually important and which are optional and which actually matter? Returning to your question whether or not we were using Pimcore Community Edition, our customer will have Enterprise starting January, will we get access to additional documentation then regarding Pimcore X upgrade?

**Comment by brusch** [comment]
I think wrapping `$clone->$name` with `try/catch` should be fine here ... so
```php
                try {
                    $clone->$name = self::loopFilterCycles($propValue);
                } catch (\Throwable $e) {
                    // just ignore errors here
                }
```

But I'm still wondering how this could happen ... a property starting with `\0` - maybe you can share some more insights on that. 

Regarding EE: there's currently no additional documentation for `pimcore/pimcore`  but you have access to support, LTS, SLA, ...

**Comment by jdreesen** [comment]
> But I'm still wondering how this could happen ... a property starting with \0 - maybe you can share some more insights on that.

A case how this could happen is e.g. if you do something like this:

```php
class Foo {
    private $foo = 'foo';
}

$foo = new Foo;
$clone = clone $foo;

foreach ((array) $foo as $prop => $val) {
    $clone->$prop = $val;
}
```

Which will result in `Fatal error: Uncaught Error: Cannot access property starting with "\0"`

This is the case, because casting an object to an array will result in the array containing *all* the object's properties, while protected ones start with `*` and private with `\0ClassName\0`:

```php
class Foo {
    private $private = 'private';
    protected $protected = 'protected';
    public $public = 'public';
}

print_r((array) new Foo);
```

```php
Array
(
    [\0Foo\0private] => private
    [*protected] => protected
    [public] => public
)
```

Here's an example: https://3v4l.org/YvSCa

**Comment by brusch** [comment]
@jdreesen aaaah 💡 Thanks for clarifying.

**Comment by michaelclaesson** [comment]
The issue with non-existing keys during preview had to do with a faulty Twig filter expression. Created a pull request that fixes this. See above.

Edit: Created the patch from the wrong branch for the bug fix initially, thus the closed first PR. Sorry! 🙈

**Comment by michaelclaesson** [comment]
I can confirm that `Fatal error: Uncaught Error: Cannot access property starting with "\0"` is an issue when trying to restore document versions as well. Added @jdreesen's fix to the serialize tool but still had an issue with `models/Element/Service.php`, more specifically the `cloneMe` method, line 1243.

Changed from this:
```
$myValue = $reflectionProperty->getValue($object);
```
To this:
```
try {
    $myValue = $reflectionProperty->getValue($object);
} catch (\Throwable $e) {
    return false;
}
```

to get it to work. No idea whether it's the right way to return false but just wanted to try whether or not it allowed me to restore and that did the job. 

What I'm curious about is when these things will be patched? Also it makes me feel a bit uncomfortable with upgrading since our client has years' worth of version data and we can't afford a situation where they can't properly restore it. Or notice at a later date that the restoration wasn't properly done. 

We're in the midst of upgrading right now and it seems like it's mostly working. Would prefer to be done before christmas. It's the version restoration and previewing that seems to be acting up. We've written a script to migrate the controller and action to the new format properly before upgrading and that takes care of having to create aliases for the old controllers. But when it comes to the version issues I feel that we're in over our heads. We simply don't have the in-house knowledge to solve these issues so we're very reliant on the product and upgrade migration working.

**Comment by dvesh3** [comment]
Fixed by #10920
